daemon/restore.go

```diff
@@ -0,0 +1,59 @@
+package daemon
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/runconfig"
+)
+
+// ContainerRestore restores the process in a container with CRIU
+func (daemon *Daemon) ContainerRestore(name string, opts *runconfig.CriuConfig, forceRestore bool) error {
+	container, err := daemon.Get(name)
+	if err != nil {
+		return err
+	}
+
+	if !forceRestore {
+		// TODO: It's possible we only want to bypass the checkpointed check,
+		// I'm not sure how this will work if the container is already running
+		if container.IsRunning() {
+			return fmt.Errorf("Container %s already running", name)
+		}
+
+		if !container.IsCheckpointed() {
+			return fmt.Errorf("Container %s is not checkpointed", name)
+		}
+	} else {
```

Comment:
For checking the image ID, do you specifically mean in the force restore case? I think one open design question is whether or not checkpointing should also cause a docker commit or otherwise create a snapshot of the filesystem. Right now, it doesn't, and so its left up to the user to figure out whether or not they need to track the filesystem. In my own system, I was concerned about quickly growing the number of images, so I opted to do my own detection of whether or not there had been a change that "mattered" before doing a commit. I'm not familiar with the content-addressability changes, so maybe that would no longer be an issue.